src/golang/cmd/server/handler/get_image_url.go

```diff
@@ -99,17 +100,73 @@ func (h *GetImageURLHandler) Perform(ctx context.Context, interfaceArgs interfac
 		return emptyResponse, http.StatusInternalServerError, errors.Wrap(err, "Unable to read container registry config from vault.")
 	}
 
-	conf, err := lib_utils.ParseECRConfig(authConf)
-	if err != nil {
-		return emptyResponse, http.StatusInternalServerError, errors.Wrap(err, "Unable to parse configuration.")
-	}
-
-	err = container_registry.ValidateECRImage(conf, args.imageName)
-	if err != nil {
-		return emptyResponse, http.StatusUnprocessableEntity, err
+	if args.service == shared.ECR {
+		conf, err := lib_utils.ParseECRConfig(authConf)
+		if err != nil {
+			return emptyResponse, http.StatusInternalServerError, errors.Wrap(err, "Unable to parse configuration.")
+		}
+
+		err = container_registry.ValidateECRImage(conf, args.imageName)
+		if err != nil {
+			return emptyResponse, http.StatusUnauthorized, err
+		}
+
+		return getImageURLResponse{
+			Url: fmt.Sprintf("%s/%s", strings.TrimPrefix(conf.ProxyEndpoint, "https://"), args.imageName),
+		}, http.StatusOK, nil
+	} else if args.service == shared.GAR {
+		conf, err := lib_utils.ParseGARConfig(authConf)
+		if err != nil {
+			return emptyResponse, http.StatusInternalServerError, errors.Wrap(err, "Unable to parse configuration.")
+		}
+
+		// Obtain an OAuth2 token
+		creds, err := google.CredentialsFromJSON(ctx, []byte(conf.ServiceAccountKey), "https://www.googleapis.com/auth/cloud-platform")
+		if err != nil {
+			return emptyResponse, http.StatusInternalServerError, errors.Wrap(err, "Unable to get credential from service account key.")
+		}
+		token, err := creds.TokenSource.Token()
+		if err != nil {
+			return emptyResponse, http.StatusInternalServerError, errors.Wrap(err, "Unable to get oauth token.")
+		}
+
+		// Create a new HTTP client
+		client := &http.Client{}
+
+		fullImageUrl := strings.Split(args.imageName, ":")[0]
+		tag := strings.Split(args.imageName, ":")[1]
+
+		host := strings.Split(fullImageUrl, "/")[0]
+		projectID := strings.Split(fullImageUrl, "/")[1]
+		repo := strings.Split(fullImageUrl, "/")[2]
+		image := strings.Split(fullImageUrl, "/")[3]
+
+		// Create a new HTTP request against the Google Artifact Registry API
+		req, err := http.NewRequest("GET", fmt.Sprintf("https://%s/v2/%s/%s/%s/manifests/%s", host, projectID, repo, image, tag), nil)
```

Comment:
Does GAR has any go library that we can use? Nowadays I'd imagine most such endpoints has go clients and shouldn't require us to implement raw http requests

Reply:
lol that's really badly maintained